### PR TITLE
BB-2-feature-calculator-service

### PR DESCRIPTION
**Pull request by MathisHTW**
This feature connects the applictation microservice with the calculator rest API microservice.
This way the application microservice can access the functionalities of the calc rest API.
This means the application api can now caluclate the VAT of a given price of a product.

### CODE REVIEW

**Comment by Ktechen** [comment]
fix Codacy Static Code Analysis issues

File: src/main/java/com/butbetter/applicationServices/Calculator.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.ResourceAccessException;
+import org.springframework.web.client.RestTemplate;
+
+@Service
+public class Calculator implements CalculatorService {
+
+    private String calcApiUrlStart = "https://caluationapi.herokuapp.com/calc/v1/VAT?price=";
```
Comment (Ktechen) [inline src/main/java/com/butbetter/applicationServices/Calculator.java:15]: I think for better readability you could write it e.g: 
```java

private String localhost = "http://localhost:8080"
private String heroku = "https://caluationapi.herokuapp.com"
private String pathOfCalutation = "/calc/v1/VAT?"
private String priceInput = "price="
private String percentInput = "&percent="
````

and than build it together

File: src/main/java/com/butbetter/applicationServices/Calculator.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        this.calcApiUrlEnd = calcApiUrlEnd;
+        log.info("urlEnd set to: " + calcApiUrlEnd);
+    }
+
+    @Override
+    public float calculateVATofPrice(float price, float percent) throws ResourceAccessException, HttpClientErrorException.BadRequest {
+        String restReq = this.calcApiUrlStart + price + this.calcApiUrlEnd + percent;
+        log.info("accessing [" + restReq + "] to calc VAT of " + price + " with VAT(" + percent + "%)");
+        CalcApiResponse calcApiResponse = calcAPIRESTTemplate.getForObject(restReq, CalcApiResponse.class);
+        if(calcApiResponse == null) throw new NullPointerException("Calc api response is null");
```
Comment (Ktechen) [inline src/main/java/com/butbetter/applicationServices/Calculator.java:48]: Is it possible to get null, if the input is only float

File: src/main/java/com/butbetter/applicationServices/CalculatorService.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package com.butbetter.applicationServices;
+
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.ResourceAccessException;
+
+public interface CalculatorService {
+
+    //public float calculateVATofProduct(Product product, float percent);
+
+    //public float calculateVATofProducts(List<Product> products, float percent);
```
Comment (Ktechen) [inline src/main/java/com/butbetter/applicationServices/CalculatorService.java:10]: Remove it

File: src/main/java/com/butbetter/applicationServices/CalculatorService.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+    float calculateVATofPrice(float price, float percent) throws ResourceAccessException, HttpClientErrorException.BadRequest;
```
Comment (Ktechen) [inline src/main/java/com/butbetter/applicationServices/CalculatorService.java:12]: Which ResourceAccessException do you mean?

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_heroku.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest
+class CalculatorTest_heroku {
+
+    @Autowired
+    private Calculator calculator;
+
+    //integration Testing on heroku:
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_heroku.java:18]: Explain me this comment

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.ResourceAccessException;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest
+class CalculatorTest_localhost {
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java:13]: This is not ruby ? Use Java Style

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_heroku.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.ResourceAccessException;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest
+class CalculatorTest_heroku {
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_heroku.java:13]: This is not ruby ? Use Java Style

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    @Test
+    @DisplayName("test calc API with the values -100 10")
+    void testCalcApiReturnForIllegalInputPrice(){
+        calculator.setCalcApiUrlStart("http://localhost:8080/calc/v1/VAT?price=");
+        assertThrows(HttpClientErrorException.BadRequest.class, ()-> calculator.calculateVATofPrice(-100,10));
+    }
+
+    @Test
+    @DisplayName("test calc API with the values 100 -10")
+    void testCalcApiReturnForIllegalInputPercent(){
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java:37]: You could remove every test.... methode

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        calculator.setCalcApiUrlStart("http://localhost:8080/calc/v1/VAT?price=");
+        assertThrows(HttpClientErrorException.BadRequest.class, ()-> calculator.calculateVATofPrice(100,-10));
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java:39]: Why are you testing the calu api ??? The Api is blackbox for you

Comment (Ktechen) [reply src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java:39]: This Api have the same tests in calculate api

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.ResourceAccessException;
+import org.springframework.web.client.RestTemplate;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SpringBootTest
+public class CalculatorTest_mockito {
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java:20]: This is not ruby ? Use Java Style

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.ResourceAccessException;
+import org.springframework.web.client.RestTemplate;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SpringBootTest
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java:19]: I think you can use @AutoConfigureMockMvc

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+public class CalculatorTest_mockito {
+
+    @Mock
+    private RestTemplate calcRestApiTemplate_Mock;
```
Comment (Ktechen) [inline src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java:23]: Naming is not ruby btw

File: pom.xml [hunk 13]
```diff
@@ -40,6 +40,21 @@
 			<artifactId>springdoc-openapi-ui</artifactId>
 			<version>1.5.2</version>
 		</dependency>
+
+		<dependency>
```
Comment (Ktechen) [inline pom.xml:44]: The dependency aren't in Stage ? Please check it

File: src/main/java/com/butbetter/applicationServices/Calculator.java (hunk 2 above)
Comment (MathisHTW) [reply src/main/java/com/butbetter/applicationServices/Calculator.java:67]: what do you mean by that?
the value of calcApiResponse can be null if there is no response by the server.
But Spring boot should throw an exception before this happens, this if clause is ther just in case something goes wrong with spring boots exception throwing.

File: src/main/java/com/butbetter/applicationServices/CalculatorService.java (hunk 4 above)
Comment (MathisHTW) [reply src/main/java/com/butbetter/applicationServices/CalculatorService.java:12]: spring boot throws a ResourceAccessException, if there is no answer from the http request.
https://stackoverflow.com/questions/45305811/throwing-resourceaccessexception-vs-httpclienterrorexception-for-resttemplate-cl

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_heroku.java (hunk 5 above)
Comment (MathisHTW) [reply src/test/java/com/butbetter/applicationServices/CalculatorTest_heroku.java:18]: i wrote 3 test classes:
1. testing calc service with localhost
2. testing calc service with restApi running on heroku
3. mocking restApi 

this comment is just to know that this class is testing the calc service with the default heroku link

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java (hunk 9 above)
Comment (MathisHTW) [reply src/test/java/com/butbetter/applicationServices/CalculatorTest_localhost.java:39]: im testing the calculator service in combination with the calc Rest API because mocking a server is tricky
i also created a test class mocking the calc Rest API but, by testing in combination i was able to cover more test cases.
This is not a Unit test obviously, but more integration/system test

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java (hunk 10 above)
Comment (MathisHTW) [reply src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java:20]: 💎

File: src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java (hunk 11 above)
Comment (MathisHTW) [reply src/test/java/com/butbetter/applicationServices/CalculatorTest_mockito.java]: i will try that looks like it could fix my issues with mocking the rest API 👍

File: pom.xml (hunk 13 above)
Comment (MathisHTW) [reply pom.xml:44]: what do you mean exactly?

File: pom.xml [hunk 14]
```diff
@@ -36,10 +36,19 @@
 		</dependency>
 
 		<dependency>
-			<groupId>org.springdoc</groupId>
```
Comment (notefox) [inline pom.xml:39]: please don't delete, add or change the the pom.xml relative to the stage version
if something should be added to it, make a branch, apply the change there and make a pull request to stage

File: src/main/java/com/butbetter/applicationservices/ApplicationServices.java [hunk 15]
```diff
@@ -1,4 +1,4 @@
-package com.butbetter.applicationServices;
```
Comment (notefox) [inline src/main/java/com/butbetter/applicationservices/ApplicationServices.java:1]: why did you change the name of the package?

File: src/test/java/com/butbetter/applicationservices/ApplicationServicesApplicationTests.java [hunk 16]
```diff
@@ -1,4 +1,4 @@
-package com.butbetter.applicationServices;
```
Comment (notefox) [inline src/test/java/com/butbetter/applicationservices/ApplicationServicesApplicationTests.java:1]: why did you change the name of the package?

File: src/test/java/com/butbetter/applicationservices/CalculatorTestUsingLocalhost.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    @DisplayName("test calc API with the values 100 -10")
+    void testCalcApiReturnForIllegalInputPercent(){
+        calculator.setCalcApiUrl("http://localhost:8080");
+        assertThrows(HttpClientErrorException.BadRequest.class, ()-> calculator.calculateVATofPrice(100,-10));
+    }
+
+    //this test asserts to true if localhost calcApi not running
+    @Test
+    @DisplayName("test not running calc API with the values 100 10")
+    void testCalcApiReturnIFCalcAPINotAvailable(){
```
Comment (notefox) [inline src/test/java/com/butbetter/applicationservices/CalculatorTestUsingLocalhost.java:45]: testCalcApiReturnIFCalcAPINotAvailable -> testCalcApiReturnIfCalcApiNotAvailable

if you use the notation Api once, then don't change
either Api or API

**Review by notefox — CHANGES_REQUESTED** [review]

File: src/main/java/com/butbetter/applicationservices/ApplicationServices.java (hunk 15 above)
Comment (MathisHTW) [reply src/main/java/com/butbetter/applicationservices/ApplicationServices.java:1]: codacy issue
